From b86fccc79574824069efbcb021f531b93909c613 Mon Sep 17 00:00:00 2001
From: Heinrich Schuchardt <xypron.glpk@gmx.de>
Date: Tue, 4 Jun 2019 17:55:32 +0200
Subject: [PATCH 1/1] efi_loader: signal events in order of decreasing TPL

UEFI events must be signaled in the order of decreasing task priority
level.

Add new events to the events linked list in decreasing TPL order.

Signed-off-by: Heinrich Schuchardt <xypron.glpk@gmx.de>
---
 lib/efi_loader/efi_boottime.c | 16 ++++++++++++++--
 1 file changed, 14 insertions(+), 2 deletions(-)

diff --git a/lib/efi_loader/efi_boottime.c b/lib/efi_loader/efi_boottime.c
index 7d1d6e9213..63145cbc35 100644
--- a/lib/efi_loader/efi_boottime.c
+++ b/lib/efi_loader/efi_boottime.c
@@ -604,7 +604,7 @@ efi_status_t efi_create_event(uint32_t type, efi_uintn_t notify_tpl,
 			      void *notify_context, efi_guid_t *group,
 			      struct efi_event **event)
 {
-	struct efi_event *evt;
+	struct efi_event *evt, *item;
 
 	if (event == NULL)
 		return EFI_INVALID_PARAMETER;
@@ -639,8 +639,20 @@ efi_status_t efi_create_event(uint32_t type, efi_uintn_t notify_tpl,
 	evt->trigger_next = -1ULL;
 	evt->is_queued = false;
 	evt->is_signaled = false;
-	list_add_tail(&evt->link, &efi_events);
 	*event = evt;
+	/*
+	 * Events must be notified in order of decreasing task priority level.
+	 * Insert the new event accordingly.
+	 */
+	list_for_each_entry(item, &efi_events, link) {
+		if (item->notify_tpl <= evt->notify_tpl) {
+			list_add_tail(&evt->link, &item->link);
+			evt = NULL;
+			break;
+		}
+	}
+	if (evt)
+		list_add_tail(&evt->link, &efi_events);
 	return EFI_SUCCESS;
 }
 
-- 
2.20.1

